### ISSUE TITLE
Abstract Python class for samplers 

### ISSUE THREAD

**Issue by gcarleo** [netket/netket#306]
In order to better integrate with other backends (pytorch, tensorflow, etc), we need to define abstract python classes for the common objects and functionalities used by the high-level python drivers of NetKet. This issue is to start a discussion about the APIs of the sampler. 
Here is a proposal, feedback is appreciated, especially in light of the requirements from other backends.  @noamwies @orsharir  

```python
from abc import ABC,property,abstractmethod

class Sampler(ABC):
    """
    Abstract base class for NetKet samplers.
    A `Sampler` generates quantum numbers $$ s = s_1\dots s_N $$ distributed
    according to the probability :

    $$P(s_1\dots s_N) = F(\Psi(s_1\dots s_N)),$$

    where F is an arbitrary function. By default F(X)=|X|^2.
    Samplers can generate multiple independent samples at once (see Sweep method).
    The current state of the sample (i.e. the value of s=s_1\dots s_N for all
    the chains) is stored in self.current_sample.
    """


    @abstractmethod
    def seed(self,base_seed):
        """Seeds the random number generator used by the Sampler."""
        pass

    @abstractmethod
    def reset(self,init_random):
        """Resets the state of the sampler, including the acceptance rate statistics
        and optionally initializing at random the visible units being sampled.

        Args:
        init_random: bool, optional
               If True the quantum numbers (visible units) are initialized at random,
               otherwise their value is preserved. (default is False)

        """
        pass

    @abstractmethod
    def sweep(self):
        """Performs a sampling sweep. Typically a single sweep
        consists of an extensive number of local moves.
        """
        pass

    @property
    @abstractmethod
    def current_sample(self):
        """Returns a 2d numpy array of currently sampled configurations. The size of
        the matrix is [N_c,N_v], where N_c is the number of independent chains being
        sampled and N_v is the size of visible vector being sampled.
        """
        pass

    @property
    @abstractmethod
    def current_state(self):
        """The current sampling state of the sampler. This contains a pair
        (visible,log_val) where log_val is the result of machine.log_val(visible)
        """
        pass

    @property
    @abstractmethod
    def n_chains(self):
        """Number of independent chains being sampled.
        """
        pass

    @property
    @abstractmethod
    def machine(self):
        """The machine (Psi(s)) used for the sampling.
        """
        pass

    @abstractmethod
    def machine_func(self):
        """function(complex): Sets and gets the function F to be used for sampling.
                       by default $$|\Psi(x)|^2$$ is sampled, however in general $$F(\Psi(v))$$
        """
        pass


```

**Comment by gcarleo** [comment]
I am currently unsure about wether ```current_state``` should be part of the interface, but for the moment I just included it.

**Comment by gcarleo** [comment]
In the specific case of the TensorFlow backend (my favorite name for that would be ```FlowKet```!) basically one would need to have autoregressive samplers etc to be derived or registered from this abstract class. @noamwies @orsharir

**Comment by noamwies** [comment]
In my opinion, the abstract sampler should be much more "abstract" and pythonic 

```python
import abc


class Sampler(abc.ABC):
    """docstring for Sampler"""

    def __init__(self, visible_size, batch_size):
        super(Sampler, self).__init__()
        self.visible_size = visible_size
        self.batch_size = batch_size

    def __iter__(self):
        return self

    @abc.abstractmethod
    def __next__(self):
        pass
```

More specifically the ```n_chains``` property is inappropriate for the non-MCMC samplers and also ```machine_func``` is debatable.

Regarding ```machine,  current_state``` we may have a helper class that will handle sampler states.

```python
class States(object):
    """docstring for States"""
    def __init__(self, machine, sampler):
        super(States, self).__init__()
        self.machine = machine
        self.sampler = sampler
    
    def __iter__(self):
        return self

    def __next__(self):
        visibles = self.sampler.__next__()
        states = self.machine(visibles)
        return visibles, states
```

The advantage of this separation  is a design that preserve the [single responsibility principle](https://en.wikipedia.org/wiki/Single_responsibility_principle)

**Comment by noamwies** [comment]
Regarding naming ```FlowKet``` is fine and better than the current ```Pyket```

**Comment by gcarleo** [comment]
1. OK, I like the idea of having an iterator type class for the sampler, and having it a bit more abstract. 
1a. ```n_chains``` is the number of independent samples generated at once by the sampler, when calling ```___next___```, thus I believe it is a rather general concept. We can change the name (maybe n_chains is not appropriate, because it refers too much to MCMC, I agree), however I feel that also ```batch_size``` is not too appropriate! For example, you might think of a sampler that generates samples using batches for efficiency, but these samples are not guaranteed to be uncorrelated. So to me, the requirement of the sampler is to generate ```sample_size``` (maybe a good name?) samples that are statistically independent. 
2a. ```machine_func``` should be improved, I agree, but that also seems necessary to me. I think that actually machine_func should just be, in the default case, F(X)=|exp(M(x))|^2, where M(x) is the actual machine call. In this way, we also remove the terminology ```log_val``` etc from the machine interface, and just leave a () operator method to compute the forward M(x). A good reason to leave F(X) inside the sampler's interface is that there are other application where we need to sample from |exp(M(X))| or other things. In that case I think it does not make sense to change the M(X) itself because, the same M(X) can be used in the same application with different purposes (say, once to sample the square, and once to sample the L^1 norm).  Notice that this would also leave the freedom to provide F(X)=|X|^2 and pass some M(X) that just computes the exp(log_val) rather than log_val (there might be cases where this is needed, I guess). Of course not all samplers might implement the possibility to change F(X)... 

2. Your ```States``` class looks like (in functionality) what is now called ```compute_samples``` see [here](https://github.com/netket/netket/blob/v2.1/netket/sampler.py). 
I agree that it should be separate from Sampler. The reason behind making ```sampler.2. __next__``` return both visible and log_val is that you'd save an extra computation to log_val (that you now perform in ```States```, since typically every sampler I know knows the value of the current log_val, while sampling it.

**Comment by noamwies** [comment]
Does general ```machine_func``` works well with returning the log_val in ```sampler.2. __next__ ``` ?

**Comment by gcarleo** [comment]
> Does general `machine_func` works well with returning the log_val in `sampler.2. __next__ ` ?

Sorry I didn't get what you mean here

**Comment by noamwies** [comment]
My meaning is that when M(x) isn't the log_val (for example in the F(X)=|X|^2 case) one should provide additional information of how to get the log_val from M(x)

**Comment by noamwies** [comment]
Regarding 2 maybe we can wrap the machine function with '''CacheManager'''?
I think this will allow good separation without hurting the performance too much, maybe even it will improve them because it can save machine evaluations for the local energy calculation

**Comment by gcarleo** [comment]
> Regarding 2 maybe we can wrap the machine function with '''CacheManager'''?
I think this will allow good separation without hurting the performance too much, maybe even it will improve them because it can save machine evaluations for the local energy calculation 

Yes, that's what I also wanted to do, that's a good idea! But I would leave caching internal to the Machine (i.e. something that machines can implement or not). Basically the way I'd do this is that if log_val(x) is computed for some x already in the cache, then you just pick it up in the some dict(x). Otherwise you compute it from scratch and put it in there. The dict is automatically reset when you assign new parameters to the machine.

**Comment by gcarleo** [comment]
> My meaning is that when M(x) isn't the log_val (for example in the F(X)=|X|^2 case) one should provide additional information of how to get the log_val from M(x)

Yes this is a bit more complicated than I though... basically I think we still need to have F(x)=|X|^2 etc like now, and keep M(x) return log_val, otherwise it's a bit complicated to get things safe for underflow and overflow when computing acceptance rates in MCMC etc. Maybe there is another way?
